### PR TITLE
Handle shutdown outside of signal handler & use context manager for progress bars

### PR DESCRIPTION
**Pull request by luator**
Handling the shutdown on SIGINT (including sys.exit()) within the signal handler function is problematic as the signal can be received multiple times before the application actually exists.  This results in the duplicated prints and potentially other issues.
Instead, only set a flag in the signal handler, which is then checked in each iteration of the main loop.  Add a simple `SignalWatcher` utility class for this.
Unfortunately this adds more redundant code in `hp_optimization()` and `grid_search()` but I don't see an easy way to avoid that at the moment.

And another change that is indirectly related as it was necessary to avoid messy output:
Use context manager for the progress bars to make sure they are properly closed when the loop finishes.  This results in much nicer output, as prints by following code will come _after_ the progress bars and not randomly somewhere _within_ them.

Below is the output of `grid_search` when run locally and Ctrl+C gets pressed.

before:
Screenshot_grid_search_before

after:
Screenshot_grid_search_after

# How I Tested
Ran examples locally.  Test on galvani is pending as the login node is not really responsive at the moment...

### CODE REVIEW

**Comment by luator** [comment]
I now tested on the MPI cluster.  Output-wise it looks good.  Jobs are still listed for a while in `condor_q` after cluster_utils terminated but since I'm also not able to kill them with `condor_rm` (saying they are not found), this might just be a glitch of condor.

**Comment by luator** [comment]
Now also successfully tested on Galvani.

**Review by mseitzer — CHANGES_REQUESTED** [review]
Very nice, thank you for tracking down the cause of this!

Eventually, SignalHandler could also be used for taking care of the slurm timeout signal. For example, we could provide a derived SignalHandler class for slurm that the user can import.

File: src/cluster_utils/server/job_manager.py [hunk 1]
```diff
@@ -361,14 +356,17 @@ def hp_optimization(
...
-        )
-
-        while cluster_interface.n_completed_jobs < number_of_samples:
+    ) as check_for_keyboard_input, redirect_stdout_to_tqdm(), SubmittedJobsBar(
+        total_jobs=number_of_samples
+    ) as submitted_bar, RunningJobsBar(
+        total_jobs=number_of_samples
+    ) as running_bar, CompletedJobsBar(
+        total_jobs=number_of_samples, minimize=minimize
+    ) as successful_jobs_bar:
```
[the remark this answers is not in the document]
Comment (luator) [reply src/cluster_utils/server/job_manager.py:365]: Ah, I didn't know ExitStack.  Will change that.

(side note: As of Python 3.9, it is possible to use parentheses for nicer multi-line formatting:
```python
    with (
        interaction_mode(cluster_interface, comm_server) as check_for_keyboard_input,
        redirect_stdout_to_tqdm(),
        SubmittedJobsBar(total_jobs=number_of_samples) as submitted_bar,
        RunningJobsBar(total_jobs=number_of_samples) as running_bar,
        CompletedJobsBar(
            total_jobs=number_of_samples, minimize=minimize
        ) as successful_jobs_bar,
    ):
```
So if we ever drop 3.8 and still remember, the ExitStack can be removed again :) )

File: src/cluster_utils/server/utils.py [hunk 2]
```diff
@@ -85,6 +86,35 @@ class ClusterRunType(enum.Enum):
     HP_OPTIMIZATION = 1
 
 
+class SignalWatcher:
+    """Watch for a signal."""
```
[the remark this answers is not in the document]
Comment (luator) [reply src/cluster_utils/server/utils.py:90]: Interesting, while working on this I at some point got the impression that Python allows multiple signal handlers (i.e. adding a new one doesn't replace the old one).  I'm not sure now, how I came to that conclusion as I just verified that it is wrong...  So thanks for pointing that out :)

File: src/cluster_utils/server/utils.py [hunk 3]
```diff
@@ -85,6 +86,35 @@ class ClusterRunType(enum.Enum):
...
+
+    def __init__(self, signal_to_watch_for: int = signal.SIGINT) -> None:
+        """
+        Args:
+            signal_to_watch_for: The signal to watch for.
+        """
+        self.received_signal = False
+        signal.signal(signal_to_watch_for, self._signal_handler)
+
+    def _signal_handler(self, sig, frame) -> None:
```
[the remark this answers is not in the document]
Comment (luator) [reply src/cluster_utils/server/utils.py:100]: I did a quick test:
```
    watcher = SignalWatcher()
    del watcher
    gc.collect()  # to make sure garbage is collected

    for _ in range(10):
        time.sleep(1)
```
When pressing Ctrl+C during the loop, it doesn't crash and the function of SignalWatcher actually still gets called (I confirmed by adding a print there).  Probably it keeps a reference somewhere for the signal handler and thus it doesn't get garbage collected.
I confirmed this by adding a `__del__` method with a print:  It doesn't get called on `del watcher` but only when registering a new signal handler.

My next thought was to register SIG_DFL in `__del__` just in case, but this is actually bad as this will get called _after_ registering a new handler.  That is
```
    watcher = SignalWatcher()
    del watcher
    signal.signal(signal.SIGINT, my_handler)
    # at this point the SIG_DFL is active, not my_handler
```

So TL;DR: It seems to be save as is, so I wouldn't try to actively do any handler cleanup.

File: src/cluster_utils/server/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (luator) [reply src/cluster_utils/server/utils.py]: I extended the docstring to cover this.

[the remark this answers is not in the document]
Comment (mseitzer) [reply src/cluster_utils/server/utils.py]: A second caveat that should be added to the docs: creating a second instance that watches the same signal as a previous SignalWatcher destroys the first one.

(A way around that would be to define a `received_signal` variable per signal type as a class variable)

File: src/cluster_utils/server/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mseitzer) [reply src/cluster_utils/server/utils.py:51]: I see. Well let's keep it as is.

File: src/cluster_utils/server/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (luator) [reply src/cluster_utils/server/utils.py]: Ah, indeed... Signal handling is so annoying :sweat:
I changed it to use a class variable now.

**Review by mseitzer — APPROVED** [review]
Looks good, thanks
